File: node/ts/io.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import * as Native from './Native.js';
import type { IoError } from './Errors.js';

/**
 * An abstract class representing an input stream of bytes.
 */
export abstract class InputStream {
  /**
   * Called to indicate the stream's resources should be released.
   *
   * The default implementation does nothing and completes immediately. Subclasses should not expect
   */
  close(): Promise<void> {
    return Promise.resolve();
  }

  /**
   * Read an amount of bytes from the input stream.
   *
   * The actual amount of bytes returned may be smaller than the amount requested by the caller, for any reason;
   * however, returning zero bytes always indicates that the end of the stream has been reached.
   *
   * @param amount The amount of bytes to read.
   * @returns A promise yielding a {@link Uint8Array} containing the read bytes.
   * @throws {IoError} If an I/O error occurred while reading from the input.
   */
  abstract read(amount: number): Promise<Uint8Array<ArrayBuffer>>;

  /**
   * Skip an amount of bytes in the input stream.
   *
   * If the requested number of bytes could not be skipped for any reason, an {@link IoError} must be raised instead.
   *
   * @param amount The amount of bytes to skip.
   * @returns A promise which is resolved once the bytes have been skipped.
   * @throws {IoError} If an I/O error occurred while skipping the bytes in the input.
   */
  abstract skip(amount: number): Promise<void>;
}

export function _bridgeInputStream(
  inputStream: InputStream
): Native.InputStream {
  return {
    read(amount: number): Promise<Uint8Array<ArrayBuffer>> {
      return inputStream.read(amount);
    },
    skip(amount: bigint): Promise<void> {
      if (
        amount < BigInt(Number.MIN_SAFE_INTEGER) ||
        amount > BigInt(Number.MAX_SAFE_INTEGER)
      ) {
        throw new RangeError('skip amount out of range');
      }
      return inputStream.skip(Number(amount));
    },
  };
}

File: rust/bridge/shared/src/io.rs
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

use std::future::Future;
use std::io;
use std::pin::Pin;

use async_trait::async_trait;

/// The result of a [`InputStream::read`].
///
/// This `enum` exists so that [`InputStream::read`] may be implemented as either a sync or async operation. In the
/// synchronous case, data is read directly into the buffer provided by the caller, to avoid the extra allocation and
/// data copy necessary in the async case.
#[must_use]
pub enum InputStreamRead<'a> {
    /// The read operation must be completed by awaiting a [`Future`] yielding the read bytes. The buffer provided to
    /// [`InputStream::read`] was not modified.
    Pending(Pin<Box<dyn Future<Output = io::Result<Vec<u8>>> + 'a>>),

    /// The read operation completed immediately, and `amount_read` bytes were copied into the buffer provided to
    /// [`InputStream::read`].
    Ready { amount_read: usize },
}

/// An input stream of bytes.
#[async_trait(?Send)]
pub trait InputStream {
    /// Read an amount of bytes from the input stream.
    ///
    /// The actual amount of bytes returned may be smaller than the buffer provided by the caller, for any reason;
    /// however, reading zero bytes always indicates that the end of the stream has been reached.
    ///
    /// # Returns
    ///
    /// An [`InputStreamRead`] indicating whether the read was completed as a sync or async operation. In the
    /// synchronous case, data is read directly into the buffer provided by the caller, to avoid the extra allocation
    /// and data copy necessary in the async case.
    ///
    /// # Errors
    ///
    /// If an I/O error occurred while reading from the input, an [`io::Error`] is returned.
    fn read<'out, 'a: 'out>(&'a self, buf: &mut [u8]) -> io::Result<InputStreamRead<'out>>
    where
        Self: 'out;

    /// Skip an amount of bytes in the input stream.
    ///
    /// # Errors
    ///
    /// If the requested number of bytes could not be skipped for any reason, including if the end of stream was
    /// reached, an error must be returned.
    async fn skip(&self, amount: u64) -> io::Result<()>;
}
